### ISSUE TITLE
Train with multi gpus in one node？

### ISSUE THREAD

**Issue by idealboy** [Tencent/tencent-ml-images#21]
sir,
Can I train with multi gpus ? Just I modify the the code in train.py

config.gpu_options.visible_device_list = str(FLAGS.train_gpus)

# FLAGS.train_gpus='0,1,2,3,4,5,6,7'

Does if affect the loss function computing or the final model's effect?

Is it a normal training process with this modified?

Thank you?

**Comment by idealboy** [comment]
Is it fittable for multi-labal train in resnet.py?

accuracy = tf.metrics.accuracy(
      tf.argmax(labels, axis=1), predictions['classes'])

**Comment by idealboy** [comment]
> @idealboy For the first question, in the provided code, we have removed the setting of multi-gpus. We may provide it in a close future.
> 
> We modify the output layer of resnet to that: there are M outouts (M is the number of trainable classes), each output has its own softmax function. Such that you can train multi-label data.

Thank you!

I just use TowerOptimizer and replicate model fn  to wrap the original optimizer and fn, so that I can use multi gpu to train! I use resnet-50, but it shows only 0.4 step/sec on 8*v100 (batchsize=512), is it normal with your oral experiment?

And what I say， you means I can train the multi label data with M binary-classifiers?

Does it affect the multi label train with keeping the train code(accuracy and predictions definitions) as it live? I just want to train it with multi-hot style.

Thank you!

**Comment by idealboy** [comment]
> > @idealboy For the first question, in the provided code, we have removed the setting of multi-gpus. We may provide it in a close future.
> > We modify the output layer of resnet to that: there are M outouts (M is the number of trainable classes), each output has its own softmax function. Such that you can train multi-label data.
> 
> Thank you!
> 
> I just use TowerOptimizer and replicate model fn to wrap the original optimizer and fn, so that I can use multi gpu to train! I use resnet-50, but it shows only 0.4 step/sec on 8*v100 (batchsize=512), is it normal with your oral experiment?
> 
> And what I say， you means I can train the multi label data with M binary-classifiers?
> 
> Does it affect the multi label train with keeping the train code(accuracy and predictions definitions) as it live? I just want to train it with multi-hot style.
> 
> Thank you!

And now, I have been trainning muli-label data with train.py(multi gpu supported modified), But it has a slow increase in train accuracy(original definition for one-hot style).

I just train with 1000 classes in openimage only.

Thank you!

**Comment by wubaoyuan** [comment]
> > > @idealboy For the first question, in the provided code, we have removed the setting of multi-gpus. We may provide it in a close future.
> > > We modify the output layer of resnet to that: there are M outouts (M is the number of trainable classes), each output has its own softmax function. Such that you can train multi-label data.
> > 
> > 
> > Thank you!
> > I just use TowerOptimizer and replicate model fn to wrap the original optimizer and fn, so that I can use multi gpu to train! I use resnet-50, but it shows only 0.4 step/sec on 8*v100 (batchsize=512), is it normal with your oral experiment?
> > And what I say， you means I can train the multi label data with M binary-classifiers?
> > Does it affect the multi label train with keeping the train code(accuracy and predictions definitions) as it live? I just want to train it with multi-hot style.
> > Thank you!
> 
> And now, I have been trainning muli-label data with train.py(multi gpu supported modified), But it has a slow increase in train accuracy(original definition for one-hot style).
> 
> I just train with 1000 classes in openimage only.
> 
> Thank you!

There are some tricks about training on such a big database. In the close future, we plan to release an arxiv paper to present the details and hyper-parameters.

**Comment by wubaoyuan** [comment]
> > > @idealboy For the first question, in the provided code, we have removed the setting of multi-gpus. We may provide it in a close future.
> > > We modify the output layer of resnet to that: there are M outouts (M is the number of trainable classes), each output has its own softmax function. Such that you can train multi-label data.
> > 
> > 
> > Thank you!
> > I just use TowerOptimizer and replicate model fn to wrap the original optimizer and fn, so that I can use multi gpu to train! I use resnet-50, but it shows only 0.4 step/sec on 8*v100 (batchsize=512), is it normal with your oral experiment?
> > And what I say， you means I can train the multi label data with M binary-classifiers?
> > Does it affect the multi label train with keeping the train code(accuracy and predictions definitions) as it live? I just want to train it with multi-hot style.
> > Thank you!
> 
> And now, I have been trainning muli-label data with train.py(multi gpu supported modified), But it has a slow increase in train accuracy(original definition for one-hot style).
> 
> I just train with 1000 classes in openimage only.
> 
> Thank you!

We have not compared the setting of M binary classifier and that of one softmax classifier. Considering the size of vocabulary, the former may be easier to train. And, the former is more flexible to handle the imbalance, which will be explained in our arxiv paper.
